sql/delete.go

```diff
@@ -18,11 +18,56 @@
 package sql
 
 import (
+	"github.com/cockroachdb/cockroach/client"
+	"github.com/cockroachdb/cockroach/proto"
 	"github.com/cockroachdb/cockroach/sql/parser"
-	"github.com/cockroachdb/cockroach/util"
 )
 
-// Delete TODO(pmattis): document.
+// Delete deletes rows from a table.
 func (p *planner) Delete(n *parser.Delete) (planNode, error) {
-	return nil, util.Errorf("TODO(pmattis): unimplemented")
+	tableDesc, err := p.getAliasedTableDesc(n.Table)
+	if err != nil {
+		return nil, err
+	}
+
+	primaryKeyPrefix := encodeIndexKeyPrefix(tableDesc.ID, tableDesc.Indexes[0].ID)
+
+	b := client.Batch{}
+
+	// TODO(tamird,pmattis): avoid going through Select to avoid encoding
+	// and decoding keys. Also, avoiding Select may provide more
+	// convenient access to index keys which we are not currently
+	// deleting.
+	node, err := p.Select(&parser.Select{Exprs: parser.SelectExprs{
+		&parser.StarExpr{TableName: tableDesc.Name},
+	}, From: parser.TableExprs{n.Table}, Where: n.Where})
+	if err != nil {
+		return nil, err
+	}
+
+	for node.Next() {
+		if err := node.Err(); err != nil {
+			return nil, err
+		}
+
+		values := node.Values()
+
+		if len(values) == 0 {
+			continue
+		}
+
+		primaryKeyValue := values[0].String()
+
+		// TODO(tamird/pmattis): delete the indexes too
+		rowStartKey := proto.Key(append(append([]byte{}, primaryKeyPrefix...), primaryKeyValue...))
```

Comment:
Yup, had this all wrong. Fixed to use `encodeIndexKey`.

On Thu, Jul 23, 2015 at 8:56 PM, Peter Mattis <notifications@github.com>
wrote:

> In sql/delete.go
> <https://github.com/cockroachdb/cockroach/pull/1790#discussion_r35387881>:
>
> > +
> > +	for node.Next() {
> > +		if err := node.Err(); err != nil {
> > +			return nil, err
> > +		}
> > +
> > +		values := node.Values()
> > +
> > +		if len(values) == 0 {
> > +			continue
> > +		}
> > +
> > +		primaryKeyValue := values[0].String()
> > +
> > +		// TODO(tamird/pmattis): delete the indexes too
> > +		rowStartKey := proto.Key(append(append([]byte{}, primaryKeyPrefix...), primaryKeyValue...))
>
> For a multi-column key you're only including the first column. That isn't
> right. This would also fail if the primary key column is not the first
> column in the table. You really need to be using encodeIndexKey here.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/1790/files#r35387881>.
>